src/Microsoft.AspNet.Routing/RouteValueDictionary.cs

```diff
@@ -243,30 +253,32 @@ public bool ContainsKey(string key)
                 throw new ArgumentNullException(nameof(array));
             }
 
+            EnsureWritable();
             ((ICollection<KeyValuePair<string, object>>)_dictionary).CopyTo(array, arrayIndex);
         }
 
         /// <inheritdoc />
-        public Dictionary<string, object>.Enumerator GetEnumerator()
+        public Enumerator GetEnumerator()
         {
-            return _dictionary.GetEnumerator();
+            return new Enumerator(this);
```

Comment:
Would it be simpler if you returned the dictionary enumerator here and used the custom enumerator if `properties` was non null?